exporter/neevafileexporter/file_exporter.go

```diff
@@ -0,0 +1,412 @@
+type neevaFileExporter struct {
+	config      *Config
+	logger      *zap.Logger
+	group       *errgroup.Group
+	spanCh      chan spanData
+	jsonCh      chan []byte
+	loopErrorCh chan struct{}
+
+	mu       sync.Mutex
+	shutdown bool // GUARDED_BY(mu)
+}
+
+type spanData struct {
+	span     pdata.Span
+	resource pdata.Resource
+}
+
+// newNeevaFileExporter creates a new exporter, after which Start and Shutdown are called.
+// The component lifecycle is described here:
+// https://github.com/neevaco/opentelemetry-collector/blob/neeva/component/component.go#L23-L33
+func newNeevaFileExporter(config *Config, logger *zap.Logger) *neevaFileExporter {
+	logger.Info("Creating neevaFileExporter",
+		zap.String("rootPath", config.RootPath),
+		zap.Duration("rotatePeriod", config.RotatePeriod),
+		zap.Int64("rotateBytes", config.RotateBytes),
+		zap.Int("json_workers", config.JSONWorkers),
+		zap.Int("json_bytes", config.JSONBytes))
+	return &neevaFileExporter{
+		config:      config,
+		logger:      logger,
+		spanCh:      make(chan spanData, config.JSONWorkers),
+		jsonCh:      make(chan []byte, 1000),
+		loopErrorCh: make(chan struct{}),
+	}
+}
+
+func (e *neevaFileExporter) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+	// Lock the whole method, to ensure we never send on e.spanCh after it's closed in Shutdown.
+	// This shouldn't be a bottleneck; we're just looping through memory and sending on spanCh.
+	// If spanCh is full and blocks, other goroutine sends would have also blocked.
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if e.shutdown {
+		return errors.New("neevaFileExporter is shut down")
+	}
+
+	resSpans := td.ResourceSpans()
+	for i := 0; i < resSpans.Len(); i++ {
+		resSpan := resSpans.At(i)
+		libSpans := resSpan.InstrumentationLibrarySpans()
+		for j := 0; j < libSpans.Len(); j++ {
+			libSpan := libSpans.At(j)
+			spans := libSpan.Spans()
+			for k := 0; k < spans.Len(); k++ {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case <-e.loopErrorCh: // avoid deadlock if the goroutine loops fail
+					return errors.New("neevaFileExporter failed with error")
+				case e.spanCh <- spanData{spans.At(k), resSpan.Resource()}:
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (e *neevaFileExporter) Start(_ context.Context, host component.Host) error {
+	// Kick off two groups of goroutines, the jsonLoops marshal spans into JSON, and the writerLoop
+	// writes and rotates files, connected with channels.  When ConsumeTraces is called it sends
+	// spans on spansCh, and when JSON marshaling is done it sends json bytes on jsonCh.
+	//
+	// Don't use the provided context arg for long-running loops:
+	// https://github.com/neevaco/opentelemetry-collector/blob/8d6bf0d2d382686eb9d21ceacdbce4b39fe51998/component/component.go#L41-L45
+	e.logger.Info("Starting neevaFileExporter")
+	ctx := context.Background()
+	e.group, ctx = errgroup.WithContext(ctx)
+	e.group.Go(func() error { return e.jsonLoops(ctx) })
```

Comment:
Optional: IIUC `jsonCh` doesn't need to be a member variable, it can be created in this function and provided to `jsonLoops` and `writerLoop` as an argument?

Reply:
Yup thanks, done.  I went back-and-forth and this is a bit cleaner.